### PR TITLE
Support for cancellation on drawables’ load

### PR DESCRIPTION
**Pull request by peppy**
This doesn't yet add support for asynchronous BDL methods, but is a first step.

### CODE REVIEW

File: osu.Framework/Graphics/Drawable.cs [hunk 1]
```diff
@@ -83,8 +83,14 @@ private void dispose(bool isDisposing)
...
                 return;
 
-            //we can't dispose if we are mid-load, else our children may get in a bad state.
-            loadTask?.Wait();
+            try
+            {
+                //we can't dispose if we are mid-load, else our children may get in a bad state.
+                loadTask?.Wait(loadTaskCancellation);
+            }
+            catch (TaskCanceledException)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Graphics/Drawable.cs]: This is in dispose tho?

File: osu.Framework/Graphics/Drawable.cs [hunk 2]
```diff
@@ -136,15 +144,17 @@ private void dispose(bool isDisposing)
...
-        internal Task LoadAsync(Game game, Drawable target, Action onLoaded = null)
+        internal Task LoadAsync(Game game, Drawable target, CancellationToken cancellation, Action onLoaded = null)
         {
             if (loadState == LoadState.NotLoaded)
             {
                 Debug.Assert(loadTask == null);
                 loadState = LoadState.Loading;
-                loadTask = Task.Factory.StartNew(() => Load(target.Clock, target.Dependencies), TaskCreationOptions.LongRunning);
+                loadTask = Task.Factory.StartNew(() => Load(target.Clock, target.Dependencies, cancellation), cancellation, TaskCreationOptions.LongRunning, TaskScheduler.Current);
+                loadTaskCancellation = cancellation;
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Graphics/Drawable.cs]: no reason, but should not cause a race as it's only used in `dispose`

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Graphics/Drawable.cs]: Fair enough

File: osu.Framework/Graphics/Drawable.cs [hunk 3]
```diff
@@ -207,10 +218,23 @@ internal void Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependen
...
                 // get our dependencies from our parent, but allow local overriding of our inherited dependency container
                 Dependencies = CreateChildDependencies(dependencies);
 
-                dependencies.Inject(this);
+                try
+                {
+                    dependencies.Inject(this);
+                }
+                catch (OperationCanceledException)
+                {
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Graphics/Drawable.cs]: yeah, this may need to be caught at a higher level..

**Comment by smoogipoo** [comment]
So if we consider something like `BeatmapInfoWedge`, which invokes `LoadComponentAsync()` whenever the beatmap changes, isn't it really weird that there's no way to cancel this without the `BeatmapInfoWedge` itself being disposed?

There should really be something like a [linked token source](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource.createlinkedtokensource?view=netframework-4.7.2), between the parent composite and the drawable to be loaded itself, somehow.
OR you should be able to provide a token to `LoadComponentAsync()` which it'll create a linked source for and the user can cancel it.

**Comment by peppy** [comment]
That's definitely a required addition. i think the linked token should be created automatically inside `LoadComponentAsync`. Would you be okay with that being a follow-up PR?

**Comment by smoogipoo** [comment]
Sure, that's fine.

**Pull request merged**